ykcapi/yk.h

```diff
@@ -51,40 +51,37 @@ void yk_location_drop(YkLocation);
 void yk_stopgap(void *addr, uintptr_t size, uintptr_t retaddr, void *rsp);
 
 #if defined(__x86_64)
-void __llvm_deoptimize(void* addr, uintptr_t size) {
+__attribute__((naked)) void __llvm_deoptimize(void* addr, uintptr_t size) {
     // Push all registers to the stack before they can be clobbered, so that we
     // can find their values after parsing in the stackmap. After that we store
     // the value of RSP in a variable so we know where to look for the register
     // values. The order in which we push the registers is equivalent to the
-    // Sys-V x86_64 ABI, which the stackmap format uses as well. However, since
-    // we are interested in the value of RSP in the previous stack frame, which
-    // has changed when we entered the __llvm_deoptimize function, we push RBP
-    // instead, which is equivalent. Note, that because of this there is no
-    // (easy) way to access RBP from the previous frame, so we just assume it's
-    // never referenced by the stackmap.
-    void *rsp;
-    __asm__ volatile (
-            "push %%rbp\n" // old RSP
-            "push %%rbp\n" // placeholder
-            "push %%rdi\n"
-            "push %%rsi\n"
-            "push %%rbx\n"
-            "push %%rcx\n"
-            "push %%rdx\n"
-            "push %%rax\n"
-            "movq %%rsp, %0\n": "=rm" (rsp));
-
-    // We can use this function's return address to lookup the correct record
-    // in the stackmap.
-    void *ret = __builtin_return_address(0);
+    // Sys-V x86_64 ABI, which the stackmap format uses as well. This function
+    // is written entirely in assembler to keep the optimiser from messing with
+    // the RSP value of the previous stackframe which is often referenced by
+    // the stackmap. For the same reason we use the "naked" attribute, which
+    // tells the compiler not to generate a prologue/epilogue.
 
-    // Forward stackmap address and size, and the return and rsp addresses to
-    // the function that parses the stackmap and initiates the stopgap
-    // interpreter.
-    yk_stopgap(addr, size, (uintptr_t) ret, rsp);
-
-    // Leave the stack as we found it.
-    __asm__ volatile ("add $0x40, %%rsp\n" :);
+    __asm__ volatile (
+        ".intel_syntax;\n"
+        // FIXME: Add other registers that may be referenced by the stackmap.
+        "push rsp;\n"
```

Comment:
I think we can add more comments to this bit, like:

 - push register contents to the stack.
 - set up the <blah> argument for the call to `yk_stopgap`.
 - free stack space.

Reply:
We might also say that rdi and rsi are passed right through.

Reply:
Is the first point not already included in the top comment? I don't think we need to explicitly say we freeing up stack space, the code makes that clear. And there's also no argument setup for `yk_stopgap` since we are just forwarding the arguments from the current function.

Reply:
> there's also no argument setup for yk_stopgap since we are just forwarding the arguments from the current function.

Lies!

```
        "mov rdx, [rsp+64];\n"
        "mov rcx, rsp;\n"
```

> Is the first point not already included in the top comment?
> I don't think we need to explicitly say we freeing up stack space, the code makes that clear.

I just wonder if we will understand what this code does a few months down the line. It's fiddly stuff, so documenting our intention may help us out a great deal later.